cardano-node/src/Cardano/Tracing/Util.hs

```diff
@@ -0,0 +1,37 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DisambiguateRecordFields #-}
+{-# LANGUAGE EmptyCase #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Cardano.Tracing.Util
+  ( fragmentLength
+  ) where
+
+import           Prelude
+import qualified Ouroboros.Network.AnchoredFragment as AF
+import           Ouroboros.Network.Block (BlockNo (..), HasHeader (..),
+                   blockNo)
+import           Ouroboros.Network.Point (fromWithOrigin)
+
+fragmentLength :: HasHeader header => AF.AnchoredFragment header -> Int
+fragmentLength af = fromIntegral $ blockN - firstBlock
+  where
+    -- Block numbers start at 1. We ignore the genesis EBB, which has block number 0.
+    blockN = unBlockNo $ fromWithOrigin (BlockNo 1) (AF.headBlockNo af)
+    firstBlock = case unBlockNo . blockNo <$> AF.last af of
+      -- Empty fragment, no blocks. We have that @blocks = 1 - 1 = 0@
+      Left _  -> 1
+      -- The oldest block is the genesis EBB with block number 0,
+      -- don't let it contribute to the number of blocks
+      Right 0 -> 1
+      Right b -> b
```

Comment:
if you use:
```
blockN = unBlockNo $ fromWithOrigin (BlockNo 0) (AF.headBlockNo af)
```
then it will simplify the calculation, you will not need a special case for `Right 0`, but you will need to add `1` which is expected: when one wants to get the number of integers `x` such that `low <= x <= high` then its `high - low + 1`.

I would renamed `blockN` as `headNo` and `firstBlock` as `lastNo`.

Reply:
This is a change in semantics for this computation, which is not something I was prepared to do.

Reply:
I see, you don't want to count the ebb, what about this:
```
headNo = unblockNo (fromOrigin (BlockNo 0) (AF.headBlockNo af)
blockLength = case unBlockNo . blockNo <$> AF.last af of
  Left _  -> headNo
  Right 0 -> headNo
  Right b -> b - headNo + 1
```
This is equivalent to your formula, isn't it?

Reply:
I don't think it's equivalent -- the minimum value produced by the existing code is 1, whereas in the above varient the minimum is 0.

I also think this discussion is what I wanted to avoid -- I just aimed to get the same logic we already have, with minimum friction.

Reply:
> I don't think it's equivalent -- the minimum value produced by the existing code is 1, whereas in the above variant the minimum is 0.
It can also be `0` because `blockNo` can be `1`. All I did I subtracted `1` from both `blockN` and `firstBlock` (which does not change the result) and inlined the subtraction, though I made a mistake it should be `headNo - b + 1` in the last case, it should be:

```
headNo = unblockNo (fromOrigin (BlockNo 0) (AF.headBlockNo af)
blockLength = case unBlockNo . blockNo <$> AF.last af of
  Left _  -> headNo
  Right 0 -> headNo -- do not count genesis EBB
  Right b -> headNo - b + 1
```

I think this is much simpler than the original.

> I also think this discussion is what I wanted to avoid -- I just aimed to get the same logic we already have, with minimum friction.
I wasn't aware of that, the diff does not show that.   If you don't want to include this in this PR, could you create another PR for that?

Reply:
I've rewritten the PR to only include minimum changes and no refactoring now.

Reply:
> I've rewritten the PR to only include minimum changes and no refactoring now. I also don't introduce semantic changes in the calculation.

I don't suggest making semantic changes, just making this cleaner whether in this PR, which makes sense, since we are adding this to `OrphanedInstances/Network`; or another one.

Reply:
@coot, I don't entirely understand your version of the code, which is why I was saying "semantic changes".

I trust you're correct, so I now took your version.

Reply:
Ok, so after further discussion with Marcin and Nicholas Frisby, it became clear that the question is clear as mud.. even if we ignore the EBB (which is a bit unfaithful in general, but is an acceptable simplification if 1) we only care about tracing and 2) in Shelley+).

Reply:
As a summary -- the semantic for the intended length should be:

- by how many blocks will be the chain elongated, as a result of adoption of the particular AF.

..or alternatively (equivalently), "how many non-anchor blocks in the AF".

Nicolas suggested going for the `AnchoredSeq` API to answer this properly.